File: Part.cs
namespace OpenModelViewFramework;

public class Part : Component
{
    // Identifier for geometry data. Size: 2 bytes.
    protected override short ID
    {
        get; set
        {
            if (value < 0 || value > short.MaxValue)
                throw new ArgumentOutOfRangeException($"Part.ID must be between 0 and {short.MaxValue} inclusive.");
        }
    }
    /*
        Color and reflection properties. Size: 4 bytes.

        0: R (0 - 255)
        1: G (0 - 255)
        2: B (0 - 255)
        3: S (0 - 255)
    */
    protected byte[] Properties
    {
        get; set
        {
            if (value == null)
                throw new ArgumentNullException("Part.Properties cannot be null.");
            else if (value.Length != 4)
                throw new ArgumentOutOfRangeException("Part.Properties length must be 4.");
        }
    }
    /*
        Data that transforms this part relative to the global coordinate system. Size: 48 bytes.
            
        r0c0, r0c1, r0c2,
        r1c0, r1c1, r1c2,
        r2c0, r2c1, r2c2,
        Δx, Δy, Δz
    */
    protected float[] Transform
    {
        get; set
        {
            if (value == null)
                throw new ArgumentNullException("Part.Transform cannot be null.");
            else if (value.Length != 12)
                throw new ArgumentOutOfRangeException("Part.Transform length must be equal to 12.");
            for (var i = 0; i < 9; i++)
            {
                if (value[i] < -1 || value[i] > 1)
                    throw new ArgumentOutOfRangeException("Part.Transform rotation elements must be between -1 and 1 inclusive.");
            }
            for (var i = 9; i < value.Length; i++)
            {
                if (value[i] == float.NegativeInfinity || value[i] == float.PositiveInfinity)
                    throw new ArgumentOutOfRangeException("Part.Transform translation elements must be between negative infinity and positive infinity exclusive.");
            }
        }
    }

    public Part(short id, bool isHidden, string path, string name, byte[] properties, float[] transform)
    {
        ID = id;
        IsHidden = isHidden;
        Path = path;
        Name = name;
        Properties = properties;
        Transform = transform;
    }
    /*
        Gets the binary representation of this part.

        data: The binary representation.
    */
    protected override void GetBinaryRep(List<byte> data)
    {
        base.GetBinaryRep(data);
        data.AddRange(Properties);
        for (var i = 0; i < Transform.Length; i++)
            data.AddRange(GetBytes(Transform[i]));
    }
    /*
        Gets the binary representation of this part.

        data: The binary representation.
        properties: The properties to get.
    */
    internal override void GetBinaryRep(List<byte> data, byte properties)
    {
        base.GetBinaryRep(data, properties);
        if ((properties & (1 << 1)) != 0)
            data.AddRange(GetBytes(ID));
        if ((properties & (1 << 2)) != 0)
            data.AddRange(Properties);
        if ((properties & (1 << 3)) != 0)
        {
            for (var i = 0; i < Transform.Length; i++)
                data.AddRange(GetBytes(Transform[i]));
        }
    }
}

File: Web/Part.js
class Part extends Component
{
    /*
        Color and reflection properties.
        
        0: R (0 - 1)
        1: G (0 - 1)
        2: B (0 - 1)
        3: S (0 - 1)
    */
    properties;
    /*
        Matrix that transforms this component relative to the global coordinate system. Assignment is column based.
        
        r0c0, r0c1, r0c2, Δx
        r1c0, r1c1, r1c2, Δy
        r2c0, r2c1, r2c2, Δz
        0, 0, 0, 1
    */
    transform;
    /*
        Collapsed transform data.

        Δx, Δy, Δz
    */
    collapsedTransform;
    /*
        Exploded transform data.

        Δx, Δy, Δz
    */
    explodedTransform;
    // Vertex array identifier.
    vertexArray;
    // Vertex buffer identifier.
    vertexBuffer;

    constructor()
    {
        super();
        this.setProperties();
        this.setTransform();
    }
    // Sets the properties using the data view.
    setProperties()
    {
        this.properties = vec4.create();
        for (let i = 0; i < 4; i++)
        {
            this.properties[i] = Component.view.getUint8(Component.index, true) / 255;
            Component.index++;
        }
    }
    // Updates the transform to use the exploded transform.
    explode()
    {
        for (let i = 0; i < this.explodedTransform.length; i++)
            this.transform[i + 12] = this.explodedTransform[i];
    }
    // Updates the transform to use the collapsed transform.
    collapse()
    {
        for (let i = 0; i < this.collapsedTransform.length; i++)
            this.transform[i + 12] = this.collapsedTransform[i];
    }
    /*
        Sets the exploded and collapsed transform of this part.

        transform: The exploded transform relative to the collapsed state.
    */
    setExplodedAndCollapsed(transform)
    {
        this.collapsedTransform = vec3.create();
        this.explodedTransform = vec3.create();
        for (let i = 0; i < this.collapsedTransform.length; i++)
        {
            this.collapsedTransform[i] = this.transform[i + 12];
            this.explodedTransform[i] = this.transform[i + 12] + transform[i];
        }
    }
    // Sets the transform using the data view.
    setTransform()
    {
        this.transform = mat4.fromValues(
            Component.view.getFloat32(Component.index, true),
            Component.view.getFloat32(Component.index + 4 * 3, true),
            Component.view.getFloat32(Component.index + 4 * 6, true),
            0,
            Component.view.getFloat32(Component.index + 4, true),
            Component.view.getFloat32(Component.index + 4 * 4, true),
            Component.view.getFloat32(Component.index + 4 * 7, true),
            0,
            Component.view.getFloat32(Component.index + 4 * 2, true),
            Component.view.getFloat32(Component.index + 4 * 5, true),
            Component.view.getFloat32(Component.index + 4 * 8, true),
            0,
            Component.view.getFloat32(Component.index + 4 * 9, true),
            Component.view.getFloat32(Component.index + 4 * 10, true),
            Component.view.getFloat32(Component.index + 4 * 11, true),
            1
        );
        Component.index += 48;
    }
    /*
        Binds the components with the given ids to the updated models in the given viewer.

        ids: The IDs to use.
        viewer: The model viewer to use.
    */
    bind(ids, viewer)
    {
        if (ids.includes(this.id))
        {
            viewer.ctx.bindBuffer(viewer.ctx.ARRAY_BUFFER, this.vertexBuffer);
            viewer.ctx.bufferData(
                viewer.ctx.ARRAY_BUFFER,
                new Float32Array(viewer.models[this.id]),
                viewer.ctx.STATIC_DRAW
            );
        }
    }
}
